internal/globalratelimit/runner/runner.go

```diff
@@ -36,62 +57,126 @@ func New(cfg *Config) *Runner {
+		r.Logger.Error(err, "failed to listen on address", "address", addr)
+		return
+	}
+	if err = r.grpc.Serve(l); err != nil {
+		r.Logger.Error(err, "failed to start grpc based xds config server")
+	}
 
-		for _, value := range snapshot.State {
-			xdsIRs = append(xdsIRs, value)
-		}
+	<-ctx.Done()
+	r.Logger.Info("grpc config server shutting down")
+	r.grpc.Stop()
+}
+
+func (r *Runner) subscribeAndTranslate(ctx context.Context) {
+	// Subscribe to resources.
+	message.HandleSubscription(r.XdsIR.Subscribe(ctx),
+		func(update message.Update[string, *ir.Xds]) {
+			r.Logger.Info("received a notification")
 
-		// Translate to ratelimit infra IR
-		result, err := r.translate(xdsIRs)
-		if err != nil {
-			r.Logger.Error(err, "failed to translate xds ir")
-		} else {
-			if result == nil {
-				r.RateLimitInfraIR.Delete(r.Name())
+			if update.Delete {
+				if err := r.addNewSnapshot(ctx, nil); err != nil {
+					r.Logger.Error(err, "failed to update the config snapshot")
+				}
 			} else {
-				// Publish ratelimit infra IR
-				r.RateLimitInfraIR.Store(r.Name(), result)
+				// Translate to ratelimit infra IR and ratelimit xDS Config.
+				rlIR, rvt, err := r.translate(update.Value)
+				if err != nil {
+					r.Logger.Error(err, "failed to translate xds ir and config")
+					return
+				}
+
+				// Update ratelimit xDS config cache.
+				if rvt != nil {
+					r.updateSnapshot(ctx, rvt.XdsResources)
+				}
+
+				// Publish ratelimit infra IR.
+				if rlIR == nil {
+					r.RateLimitInfraIR.Delete(r.Name())
+				} else {
+					r.RateLimitInfraIR.Store(r.Name(), rlIR)
+				}
 			}
-		}
-	}
+		},
+	)
 	r.Logger.Info("subscriber shutting down")
 }
 
-func (r *Runner) translate(xdsIRs []*ir.Xds) (*ir.RateLimitInfra, error) {
+func (r *Runner) translate(xdsIR *ir.Xds) (*ir.RateLimitInfra, *types.ResourceVersionTable, error) {
 	ratelimitInfra := new(ir.RateLimitInfra)
+	resourceVT := new(types.ResourceVersionTable)
 
-	for _, xdsIR := range xdsIRs {
-		for _, listener := range xdsIR.HTTP {
-			cfg := translator.BuildRateLimitServiceConfig(listener)
-			if cfg != nil {
-				str, err := translator.GetRateLimitServiceConfigStr(cfg)
-				if err != nil {
-					return nil, fmt.Errorf("failed to get rate limit config string: %w", err)
-				}
-				c := &ir.RateLimitServiceConfig{
-					Name:   listener.Name,
-					Config: str,
-				}
-				ratelimitInfra.ServiceConfigs = append(ratelimitInfra.ServiceConfigs, c)
+	for _, listener := range xdsIR.HTTP {
+		cfg := translator.BuildRateLimitServiceConfig(listener)
+		if cfg != nil {
+			// Add to xDS Config resources.
+			resourceVT.AddXdsResource(resourcev3.RateLimitConfigType, cfg)
+
+			str, err := translator.GetRateLimitServiceConfigStr(cfg)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to get rate limit config string: %w", err)
 			}
+			c := &ir.RateLimitServiceConfig{
+				Name:   listener.Name,
+				Config: str,
+			}
+			ratelimitInfra.ServiceConfigs = append(ratelimitInfra.ServiceConfigs, c)
```

Comment:
curious why is the service config needed now that we are using xds ?

Reply:
because for `ratelimitInfra`, we only use it to update the configMap `envoy-ratelimit`:

https://github.com/envoyproxy/gateway/blob/a9a85281f03cada583f65dca61c111e7c36e10d9/internal/infrastructure/kubernetes/ratelimit/resource_provider.go#L48-L50

that configMap `envoy-ratelimit` is used when we run ratelimit provider in FILE mode.

like the discuss above, someone may want to override the XDS mode to FILE manually. so if people does that, we can still make sure that ratelimit provider is working.

Reply:
imo we should remove all the configmap logic, its more maintenance burden  for this project to support multiple config methods. afaik there is no use case to support both, users would like to configure ratelimiting and are not concerned with how ratelimiting is being implemented

Reply:
agree, I've removed the `FILE` mode.